File: lang/python/test/muse/data_structures/ArrayListTest.py
from muse.data_structures.ArrayList import ArrayList
from muse.util import TestRunner

def testArrayList() -> bool:
    size = 8192

    list = ArrayList(0)

    for i in range(1, size + 1):
        list.append(i)

    list.shrink()

    if list.size() != size:
        return False

    if list.capacity() != size:
        return False

    for i in range(size):
        if list.get(i) != i + 1:
            return False

    for i in range(size):
        list.set(i, size - i)

    for i in range(size):
        if list.get(i) != size - i:
            return False

    x = 0
    y = 0

    i, j = 0, size - 1
    while i < j:
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
        i, j = i + 1, j - 1

    for i in range(size):
        if list.get(i) != i + 1:
            return False

    for i in range(size, 0, -1):
        if list.back() != i:
            return False
        list.eject()

    list.shrink()

    if not list.isEmpty():
        return False

    return list.capacity() == 0

def main() -> None:
    TestRunner.parseTest(testArrayList())

if __name__ == "__main__":
    main()

File: go/muse/data_structures/ArrayList/ArrayList_test.go
package ArrayList

import "testing"

func TestArrayList(t *testing.T) {
	size := 8192

	list := New(0)

	for i := 1; i <= size; i++ {
		list.Append(i)
	}

	list.Shrink()

	if list.Size() != size {
		t.FailNow()
	}

	if list.Capacity() != size {
		t.FailNow()
	}

	for i := 0; i < size; i++ {
		if list.Get(i) != i+1 {
			t.FailNow()
		}
	}

	for i := 0; i < size; i++ {
		list.Set(i, size-i)
	}

	for i := 0; i < size; i++ {
		if list.Get(i) != size-i {
			t.FailNow()
		}
	}

	for i, j := 0, size-1; i < j; i, j = i+1, j-1 {
		x := list.Get(i)
		y := list.Get(j)

		list.Remove(i)
		list.Insert(i, y)
		list.Remove(j)
		list.Insert(j, x)
	}

	for i := 0; i < size; i++ {
		if list.Get(i) != i+1 {
			t.FailNow()
		}
	}

	for i := size; i >= 1; i-- {
		if list.Back() != i {
			t.FailNow()
		}
		list.Eject()
	}

	list.Shrink()

	if !list.IsEmpty() {
		t.FailNow()
	}

	if list.Capacity() != 0 {
		t.FailNow()
	}
}
